### ISSUE TITLE
New interactive editor in lesson is not accessible for those using screen readers

### ISSUE THREAD

**Issue by jdwilkin4** [freeCodeCamp/freeCodeCamp#62895]
We had a now deleted post on reddit about the difficulty using the new interactive editor in the lessons. 

Here is an example lesson

https://www.freecodecamp.org/learn/full-stack-developer/lecture-understanding-html-attributes/what-is-html

The post was saying they can no longer just read straight down the page and it is causing them to pause and readjust around the interactive examples. 

We will need to hide the code examples from assistive technology.

Also, we should probably do more accessibility testing to make sure there are no other issues and resolve them.

**Comment by jdwilkin4** [comment]
I moving this to discussing because hiding the editor is part of the problem. Once it is hidden, those using screen readers won't be able to see the examples at all. 
So there needs to be further discussion on how to best show examples so campers have the option to use the interactive examples but also those who can't will still be able to see the examples in some way.

**Comment by ojeytonwilliams** [comment]
Given that everyone seems to get on reasonably well with the old challenge editor, it might be enough to use the Monaco editor instead of Codemirror. After all, we originally moved from Codemirror to Monaco since we wanted a more accessible editor.

**Comment by ojeytonwilliams** [comment]
I hadn't understood the problem correctly. In case anyone else makes the same mistake, the problem is not the accessibility of the interactive editors, exactly, it's that they interrupt the reading experience. It's a pain to navigate the page since it's continually be interrupted by the editors and you have to work around them. Previously, when they were simply code examples, you could listen to the entire page without any need to jump around.

**Comment by huyenltnguyen** [comment]
I'm thinking out loud here:
- We can display code examples using `PrismFormatted` and hide them visually with `class="sr-only"` as a replacement for the interactive editor. 
  - This would support blind and low-vision users, but could confuse sighted screen reader users since the visual and auditory content wouldn't align.
- Alternatively, we could show both the interactive editor and static code blocks in tabs, and save the user's preference in local storage so their selected tab opens by default. 
  - Though, this doesn't address the page navigation issue since we're even introducing more interactive elements between paragraphs, which isn't great for users relying on keyboard navigation. 
  - UI-wise, the tab display can also be distracting, as sighted users mostly likely prefer seeing the interactive editor.
- Another approach is to add an "Interactive Code Examples" option to the settings page, similar to the "Keyboard Shortcuts" option, so users can choose (and maybe we turn it off by default).
  - This does require users to visit `/settings`, so the UX isn't ideal.
  - Side note: Even if we don't go with this approach, it might be worth adding an "Accessibility" section to the settings page at some point (like https://github.com/settings/accessibility), so users have more control over what they see / how they use our platform.
- Since coding challenges have a toolbar/action bar at the top, we could add the same bar to these challenge types and include a button for toggling interactive code examples.
    <details>
      <summary>Screenshot</summary>
	  
    <img width="1437" height="405" alt="Image" src="https://github.com/user-attachments/assets/df04f8cd-5d3e-4256-81c9-453abefc6940" />

    </details> 

  - Other potential uses for this action bar:
      - A button for showing the content outline (https://github.com/freeCodeCamp/freeCodeCamp/issues/61643)
      - A transcript toggle on language lessons (related to https://github.com/freeCodeCamp/freeCodeCamp/pull/62492) 
        - For context: we added a transcript button to language lessons for a11y support, but we don't want to encourage users to click it since they're supposed to listen to the audio. One of my suggestions was to move the Transcript button elsewhere, away from the Play button and higher up the page (because users don't usually reach that far). I discussed with Nielda and we decided to disable persistence as a stopgap, until we can revisit the UI/UX of language lessons.
  - Semi-related: As I was implementing #61643 (which is ~abandoned~ still a WIP), I ran into an issue managing the sidebar height because the breadcrumbs aren't sticky and disappear as the user scrolls down. I was going to make the breadcrumbs sticky to make my life easier. But that change actually makes sense, since on long pages like quizzes or reviews, users currently have to scroll all the way up to find them. So if we are adding a toolbar, both it and the breadcrumbs should be sticky.

---

So, I'm leaning toward the toolbar approach.

And regardless of the UI approach, we would still need to update the challenge parser to include the raw HTML of the code blocks in `nodule` data.

**Comment by ojeytonwilliams** [comment]
Thanks Huyen, I'm not the best person to judge, but both those ideas sound promising.

In the short term, how about going with `sr-only` code examples while we work on the full solution? If that sounds reasonable, I'll put together a quick PR to adjust the parser.

**Comment by jdwilkin4** [comment]
> We can display code examples using PrismFormatted and hide them visually with class="sr-only" as a replacement for the interactive editor.

I am not sure I understand this one. This sounds like a static code example that we had before without the preview result. If that is the case, then that creates the issue we had before where campers can't see the output of the code.

**Comment by ojeytonwilliams** [comment]
Yep, this wouldn't be a good solution, but it would only be live while we come up with something acceptable.

To my mind it's less frustrating to have a description of code you can't interact with than a gap where the code/editor would be. For instance, a lot of the lessons read like this "For example the code below has...". If that code is hidden from you the lesson is basically useless.

**Comment by jdwilkin4** [comment]
ok, we will go with the sr-only code examples for now until the toolbar approach is ready. 👍🏾

**Comment by huyenltnguyen** [comment]
Sorry I wasn't clear about the `PrismFormatted` and `sr-only` setup.

The idea is to provide a tailored experience to each user group:
- Sighted users interact with the editor, which is hidden from screen readers.
- Non-sighted users access the same content through the `PrismFormatted` code blocks, which are hidden from sighted users.

**Comment by jdwilkin4** [comment]
ahhh I see. Thanks for the clarification.

Then we should be fine moving forward with the `PrismFormatted` approach. 

Does that mean we don't need the toolbar solution?

**Comment by ojeytonwilliams** [comment]
I think we'd still need something else, since it's not truly accessible if part of the experience isn't available for all users.

**Comment by huyenltnguyen** [comment]
So the PrismFormatted approach is just one of the options. It's not the best as we need to consider sighted users who use screen readers. This approach would confuse them as the thing they see on the screen (interactive editor) doesn't match what the screen reader announces (the PrismFormatted code).

The best option is probably:
- Showing the toolbar
- Add an "Interactive editor" button to the toolbar
- Maybe keep it off by default and show the static code examples (PrismFormatted **without** `sr-only`)
- Persist the selected mode

**Comment by jdwilkin4** [comment]
ok, we can move forward with the toolbar approach for the long term solution. 

Since this sounds like it will probably be a few months of work and reviews before this change is live, what do we do in the mean time?

Do we keep adding interactive examples and keep it hidden for screen readers? Or do we need to do the PrismFormatted with sr-only now as a temporary solution?

**Comment by huyenltnguyen** [comment]
The toolbar should be a straight forward change so I can work on this now, unless Oliver has already started (@ojeytonwilliams do shout if you have). Or one of us can also implement the PrismFormatted + sr-only approach in parallel, in case the toolbar turns out to be challenging.

We can continue with #61859 in the meantime.

**Comment by ojeytonwilliams** [comment]
Sorry, Huyen, I thought the real fix was going to take a while. If you reckon it's pretty straightforward, please go ahead.

Let me know if I can help.

**Comment by huyenltnguyen** [comment]
It was my bad. So sorry for the confusion there, I wasn't in the right headspace so my communication was a mess 😅 

I've got a draft for the functionality, and it just needs some CSS adjustments, which I'll get to when I'm up.

**Comment by ojeytonwilliams** [comment]
It's all good, thanks for taking care of that.

**Comment by QuincyLarson** [comment]
Thanks @huyenltnguyen for implementing a solution. We do want to keep the UI as simple as possible with as a few buttons as absolutely necessary. If it's possible to create hotkeys that low vision learners can use (instead of adding a bunch of buttons that aren't relevant for fully sighted learners) then we should do that instead, and just make sure low vision learners are aware of the hotkeys.

**Comment by raisedadead** [comment]
We may have tried this but https://sandpack.codesandbox.io/docs/guides/integrate-monaco-editor lets users add monaco, which is a11y compatible?

**Comment by ojeytonwilliams** [comment]
@raisedadead that's a separate issue, I think. We should do that so that the editor is more accessible, but the problem here is that the editor + preview component is breaking the reading flow.

**Comment by raisedadead** [comment]
Well it would kill two birds with one stone wouldn't it? Why have two implementations at all?

**Comment by ojeytonwilliams** [comment]
I don't think so, but we want the more accessible editor anyway, so I'll see what difference it makes.

Update: I did a quick test and it's still stopping the flow. The screenreader will read up to the interactive element and then pause for input. As I understand it, that's the essence of the original complaint.

**Comment by raisedadead** [comment]
Okay, created a dedicated issue.
